### PR TITLE
Add new "site search" page

### PR DESCRIPTION
**Pull request by cpeel**
This adds a new "site search" page. Users can either click on the search icon in the header or use the `/` hotkey in any non-input field on a site page. This has been available for public preview & discussion in the [New site search feature](https://www.pgdp.net/phpBB3/viewtopic.php?f=4&t=81458) forum.

Sandbox: https://www.pgdp.org/~cpeel/c.branch/site-search/

### CODE REVIEW

File: tools/site_search.php
Comment (bpfoley) [inline tools/site_search.php:153]: Isn't `$prefix_searches[$key]` just `$data` for the various occurrences this loop?

I tried it out and I think you can write this a bit more effectively as the follow (note the use of a reference in the foreach)

```
  foreach($prefix_searches as $key => &$data) {
      if (($alias = @$data["alias"]) !== null) {
          unset($data["alias"]);
          $prefix_searches[$alias] = $data;
      }
  }
```

File: tools/site_search.php
Comment (bpfoley) [inline tools/site_search.php:251]: as above...

File: tools/site_search.php
Comment (bpfoley) [inline tools/site_search.php:321]: Again, I'm probably being fussy here, but we have a mild case of 'arrow code' here. We could avoid the nested indentation with 
```
if (preg_match('/^(\w+):\s*(.*)$/', $query, $matches) === false) return;
...
$data = ...
if ($data == null) return;
...
$query = ...
metarefresh(...);
```

File: tools/site_search.php
Comment (srjfoo) [inline tools/site_search.php:228]: Is the blog supposed to point to the wiki?

**Comment by srjfoo** [comment]
I don't know if this has the potential for being a problem or not, but in testing, I've found myself using the `/` from the help page, which, of course, has the form input box highlighted and active. Which, of course, breaks whatever I'm searching for. Would it be reasonable to clear the form if a `/` is typed (including being cleared of the `/`? Or maybe the `/` should redirect to the form in the header?

For that matter, once the search term is entered, if it's not a valid search, the cursor moves to the beginning of the search string that's still in the form. Should it be being cleared if it's an invalid search term? I think what I'm getting at is, if the user does get an invalid search response, is there a way to get out of it without using the mouse other than editing the search field, or is that just one of the drawbacks?

Observations:
* The wiki search is just a basic wiki search, it looks like. If you search for a username, pages that have that username on them will come up, but the user's personal page will not (I tried adding `User` namespace search and making it persistent, but that didn't work). As I said, just an observation; I don't know if it's worth worrying about.
* Prefix searches are exact string, not substring (e.g. `forum: test` does not also give results for `tests` or `testing`).

File: tools/site_search.php
Comment (cpeel) [reply tools/site_search.php:228]: No, I just caught this myself!

**Comment by cpeel** [comment]
For failed searches, my inclination was to put the failed search terms in the box in case someone wanted to edit them and search again. If someone fat-fingers something like searching for `titl:mark twain` would they rather add the `e` or just type the whole thing over again? I'm not stuck on this and we could just show an empty input field, but that was my rationale.

We could prevent the search box on the page from having focus. That would make `/` work on the page again when it loads.

We could do a combination of either of the above.

I don't want to do anything special for `/` within that search box because that would be extremely unexpected -- when you're in an input field you expect text to be input, not have it not show up or have it move you outside the focus of the field.

> The wiki search is just a basic wiki search

Good call. I just made it do a full search which I think will be better.

> Prefix searches are exact string, not substring

Correct. This was by design -- the full string or the alias works, but not a subset.

---

I also meant to add a text to let users know they can use `/` to bring up the search field on any page. I'll see about getting that in there now.

**Comment by cpeel** [comment]
I've addressed most of the PR comments (thanks @bpfoley) and added a tip on the use of `/` at the bottom.

After thinking about the auto-focus and failed search terms in the box, what I've decided is that we should think about 2 primary users:
* keyboard users probably want to hit `/` on the page and get back to searching without using the mouse. They're not going to want to correct the failed search anyways probably.
* mouse users will probably want to click in the box to edit their failed terms.

So I've removed the auto-focus from the page in hopes that it works for both users.

**Comment by srjfoo** [comment]
> For failed searches, my inclination was to put the failed search terms in the box in case someone wanted to edit them and search again. If someone fat-fingers something like searching for `titl:mark twain` would they rather add the `e` or just type the whole thing over again? I'm not stuck on this and we could just show an empty input field, but that was my rationale.

Perfectly reasonable thinking.

> We could prevent the search box on the page from having focus. That would make `/` work on the page again when it loads.

I like the idea of removing focus on the search box on a failed search so that the `/` works. Would that make it harder for those who want to edit the input on the help page?

> We could do a combination of either of the above.

How would a combination work?

> I don't want to do anything special for `/` within that search box because that would be extremely unexpected -- when you're in an input field you expect text to be input, not have it not show up or have it move you outside the focus of the field.

I was really pretty much just tossing stuff at the wall in hopes that something might be helpful 😁 

> > The wiki search is just a basic wiki search
> 
> Good call. I just made it do a full search which I think will be better.

I really like that -- it should make it easier, for instance, to search for a template. Or a file. (Initially, when I tried the updated version, I did it without the `:`, and nearly said "butbutbut it's not working! 😆 )

> > Prefix searches are exact string, not substring
> 
> Correct. This was by design -- the full string or the alias works, but not a subset.

That's fine -- I just wanted to make sure it was intentional.
 
> I also meant to add a text to let users know they can use `/` to bring up the search field on any page. I'll see about getting that in there now.

👍

**Review by 70ray — COMMENTED** [review]
Noticed two minor things while testing:

1: If you put in something that it doesn't understand (like ab) you get a warning message. Then do a sensible search (like ah). Then go back to the previous page, you get to the page with the warning message. It seems like it would be better to get back to the raw page or have 'ah' filled in.

2: If you search for a language without capitalizing the first letter, the search understands the language correctly but if you then then do a refine search, the search does not remember the language.

**Comment by cpeel** [comment]
> 1: If you put in something that it doesn't understand (like ab) you get a warning message. Then do a sensible search (like ah). Then go back to the previous page, you get to the page with the warning message. It seems like it would be better to get back to the raw page or have 'ah' filled in.

I agree this is not ideal. To do that we would need to not ever put the failed search term in the search box and at least right now I think that's the better trade-off. It's quite possible that I'm wrong and we should just never put the failed search term in the box and if the consensus (between the devs in this PR or users post-deploy) is that we don't then I'm happy to change that.

> 2: If you search for a language without capitalizing the first letter, the search understands the language correctly but if you then then do a refine search, the search does not remember the language.

Good find. I don't see an easy way to fix this without doing a fair amount of introspection on the data on the site search page itself. Right now you can search for parts of languages "eng", "english", "engli" and they all work to search, but will all fail if you refine it. I think we should either accept this limitation or remove `language` as something you can search on with the site search (which is my preference as a project language isn't going to be highly selective). @bpfoley you requested this feature, what do you think?

**Comment by bpfoley** [comment]
Hmm. As you say, the selectivity of language on its own is pretty low and I'd imagine much of the time language specialists will also work in a specific round, so you really want to be able to search for, e.g. Hawaiian and F1. If refine search isn't working and is a lot of extra work to fix, I'd say it's probably not worth having language at all...

Sorry for creating the extra churn.

**Comment by windymilla** [comment]
> Good find. I don't see an easy way to fix this without doing a fair amount of introspection on the data on the site search page itself. Right now you can search for parts of languages "eng", "english", "engli" and they all work to search, but will all fail if you refine it. I think we should either accept this limitation or remove `language` as something you can search on with the site search (which is my preference as a project language isn't going to be highly selective). @bpfoley you requested this feature, what do you think?

If we decide to keep language, I don't think that limitation is too serious, but I agree language is less useful than the others, so I'd vote to remove it. 
I suppose you could consider it a problem with the Project Search/Refine page that you can give it a URL that it can't reflect in the search form, e.g. `https://www.pgdp.org/~cpeel/c.branch/site-search/tools/search.php?show=search&language[]=f` lists all projects that have `f` in the language, but the search form can't reflect that. However, I doubt that it's something the users will notice or need. A workaround is to just re-select the language when using the form to refine.

**Review by windymilla — APPROVED** [review]
Excellent feature!

**Comment by srjfoo** [comment]
> > 1: If you put in something that it doesn't understand (like ab) you get a warning message. Then do a sensible search (like ah). Then go back to the previous page, you get to the page with the warning message. It seems like it would be better to get back to the raw page or have 'ah' filled in.
> 
> I agree this is not ideal. To do that we would need to not ever put the failed search term in the search box and at least right now I think that's the better trade-off. It's quite possible that I'm wrong and we should just never put the failed search term in the box and if the consensus (between the devs in this PR or users post-deploy) is that we don't then I'm happy to change that.

The first thing that came to mind to test this on other sites was archive.org. I tried their advanced search (which I know uses keywords) and set up a search that I knew would work, then changed it slightly to get a null result. Then fixed it again. Going back to the previous page takes you back to the unsuccessful result, so I don't think that's the wrong approach.
> 
> > 2: If you search for a language without capitalizing the first letter, the search understands the language correctly but if you then then do a refine search, the search does not remember the language.
> 
> Good find. I don't see an easy way to fix this without doing a fair amount of introspection on the data on the site search page itself. Right now you can search for parts of languages "eng", "english", "engli" and they all work to search, but will all fail if you refine it. I think we should either accept this limitation or remove `language` as something you can search on with the site search (which is my preference as a project language isn't going to be highly selective). @bpfoley you requested this feature, what do you think?

I played with this, too. If you do a title search, and refine the search, the text you typed in *is* in the project search form. I assume that the difference is that title (also author) is a text field, and language is a multi-select?

That being said, I agree with you, Brian and Nigel. I'd just remove it. I would think that language would be best used in conjunction with something else (like all German projects available for PP, for instance) and that's just easiest done from the search form itself.

`<tb>`

Rambling thoughts: As I was doing this round of testing it occurred to me to ask if we should allow a search on PG number (which would point to the project page for that project). Unlike searching for a project by projectID, that would require a keyword. I don't feel strongly about it, though. I would not even expect a search on projectid to be generally useful unless someone is working with a list of bare projectIDs, and  (for those of us who might be working from a list of bare projectIDs, yes. But even then, you're better off copying the list of projectids and pasting it into a project search form, because you can give that form multiple projectIDs instead of doing them one at a time.

File: pinc/theme.inc
Comment (chrismiceli) [inline pinc/theme.inc:184]: The current HTML output does not change the mouse on hover, so I wasn't aware this search icon was interact-able at first. Could we make the cursor a hand on hover link an anchor?

File: tools/site_search.php
Comment (chrismiceli) [inline tools/site_search.php:303]: can a user name match a project id? Same for can a user name be one of our jumps?

File: tools/site_search.php
Comment (chrismiceli) [inline tools/site_search.php:329]: *minor* is that space after `!` normal?

File: pinc/theme.inc
Comment (cpeel) [reply pinc/theme.inc:184]: Good catch - I'll fix this.

File: tools/site_search.php
Comment (cpeel) [reply tools/site_search.php:303]: Technically a username could match a project ID and in that case the requester would be taken to the project, not the user. They could use the `user:` prefix to specify they want the user not the project.

Most all of our jump words would be valid usernames. The jump words are evaluated first for this reason. See also [this line](https://github.com/DistributedProofreaders/dproofreaders/pull/1184/files#diff-05373f93c16a82c90bebb30d9ec57edeaf43a5e3d8e35aae34d1f1ad2b59ab21R18).

File: tools/site_search.php
Comment (cpeel) [reply tools/site_search.php:329]: Looks like most of our code uses no-space, but we have spaces elsewhere, eg:
```
user_is.inc:        if (!$shown) {
user_is.inc:    if (! is_numeric($limit)) {
```
Sad that `php-cs-fixer` isn't standardizing those. I'll fix this one to match the bulk of our code.

**Comment by cpeel** [comment]
`language:` prefix removed, "Search" name is now also a link and both the icon and the name use pointers for cursors. Addressed other PR feedback.

**Comment by srjfoo** [comment]
See the image below. The pointer seems to work for the search icon with no problem, and I tried both French and German, and while I was able to keep the pointer from working on the others occasionally, it's not so bad, but "Mitteilungen" is long enough that there's a lot of empty space within the rectangle that shows as a pointer, but doesn't actually bring up the link, both to the left, where the pointer in the image, below, is, and to the right between the number and the label (I assume that it would mirror the left if there weren't a number there.

image 

Looking more closely (literally, zooming in on the page header), and clicking around all of the icons (back in English), I'm seeing some inconsistency. I couldn't get the pointer with no link for the wiki or blog icons, but I could for Inbox, Forums and Help. The search icon is a bit different, and clicking anywhere in the enclosing rectangle triggers it just fine.

(I apologize for the image quality, but while I have great fun doing screenshots in general, I can't seem to make them show the cursor, even with that option turned on, so I took a pic with my phone to get this.)

**Comment by 70ray** [comment]
> I suppose you could consider it a problem with the Project Search/Refine page that you can give it a URL that it can't reflect in the search form, e.g. `https://www.pgdp.org/~cpeel/c.branch/site-search/tools/search.php?show=search&language[]=f` lists all projects that have `f` in the language, but the search form can't reflect that. However, I doubt that it's something the users will notice or need. A workaround is to just re-select the language when using the form to refine.

Yes, this is really a problem with the Project Search page, possibly fixable later.

**Comment by chrismiceli** [comment]
Minor UX feedback, the error message doesn't stand out that much. With most of the page being red, it was difficult to see what went wrong
image

File: pinc/theme.inc
Comment (cpeel) [reply pinc/theme.inc:217]: Yeah, I will pull out the `help` into an `sprintf()` so that doesn't get translated.

**Comment by cpeel** [comment]
> Minor UX feedback, the error message doesn't stand out that much. With most of the page being red, it was difficult to see what went wrong

These are using the existing `warning` style for paragraphs and our standard style for `<kbd>` tags to indicate keyboard input. We could style the "don't know what to do with the search" text differently (`error` type seems a little dramatic for this -- neither `warning` or `error` seem entirely accurate) and/or we could change the `<kbd>` style across the site. The former is probably easier to do as I don't really want to revisit the `<kbd>` style across the site. Thoughts?

**Comment by cpeel** [comment]
> Looking more closely (literally, zooming in on the page header), and clicking around all of the icons (back in English), I'm seeing some inconsistency.

I've revised the "how things link" and styling once again and now the icons and texts are links and as links are the only thing that show the pointer icon.

I hope this also resolves @chrismiceli 's comment about breaking tabbing as now everything in the header -- including search -- is a link.

> Yes, this is really a problem with the Project Search page, possibly fixable later. So I think we should keep the language option.

We can re-add the language option if/when we fix the project search. Easier to add things we know work than release things that don't quite work and field questions.

File: pinc/theme.inc
Comment (srjfoo) [reply pinc/theme.inc:217]: I would expect all keywords to be keywords. It's true, I haven't checked any other sites that use keyword searches to see if they translate them, but, I think that it's reasonable not to have translated versions. The tooltips should be translated, though.

Which makes me wonder -- for the help message, should a TRANSLATORS note be added to suggest that a comment be added to the translated version that the keywords work only as given?

Comment (windymilla) [reply pinc/theme.inc:217]: I'm not suggesting it should be changed now, but just wondered about it. I did a quick trial in Slack. You can type `/open #general` for example to open a channel named general. If you change your language to French, you can type either `/open #general` or `/ouvrir #general`. So, it seems as though the French version of the keyword is added, but the English version is also retained.

Comment (cpeel) [reply pinc/theme.inc:217]: Interesting! I think we should start with just English. Then we can work with the translators on custom prefix & jump words. I don't think just wrapping them in `_()` is going to be the right approach here.

Comment (windymilla) [reply pinc/theme.inc:217]: Definitely agree!

Comment (srjfoo) [reply pinc/theme.inc:217]: > Interesting! I think we should start with just English. Then we can work with the translators on custom prefix & jump words. I don't think just wrapping them in `_()` is going to be the right approach here.

I agree. Somehow, you'd have to link the translation in the PO fle to the keyword, wouldn't you?

**Review by 70ray — APPROVED** [review]

**Comment by cpeel** [comment]
> I don't suppose there's a way to get rid of the latter? It's not really needed.

We're not doing anything "special", that's just the browser indicating that the input field has focus. I don't know of anything that could change it but my CSS-foo is not amazing.

**Comment by srjfoo** [comment]
> > I don't suppose there's a way to get rid of the latter? It's not really needed.
> 
> We're not doing anything "special", that's just the browser indicating that the input field has focus. I don't know of anything that could change it but my CSS-foo is not amazing.

Based on what other sites do, there is definitely a way to get rid of it (for instance this text input box has a blue highlight while the field is active; phpbb on PMs of forum posts has no special highlighting at all. But like you, I have no idea how to do it. It's not the end of the world, just a bit more emphasis than I think the field really needs. And I just checked -- it's not nearly as much in-your-face on the lighter themes. 🤣

**Review by chrismiceli — APPROVED** [review]
It could be cool if we could autocomplete some of the jump options? Definitely a nice to have https://jsfiddle.net/7jdzntke/

**Comment by bpfoley** [comment]
> It could be cool if we could autocomplete some of the jump options? Definitely a nice to have https://jsfiddle.net/7jdzntke/

Agreed. I was thinking this too, but forgot to bring it up.

**Comment by 70ray** [comment]
> > > I don't suppose there's a way to get rid of the latter? It's not really needed.
> > 
> > 
> > We're not doing anything "special", that's just the browser indicating that the input field has focus. I don't know of anything that could change it but my CSS-foo is not amazing.
> 
> Based on what other sites do, there is definitely a way to get rid of it (for instance this text input box has a blue highlight while the field is active; phpbb on PMs of forum posts has no special highlighting at all. But like you, I have no idea how to do it. It's not the end of the world, just a bit more emphasis than I think the field really needs. And I just checked -- it's not nearly as much in-your-face on the lighter themes. 🤣

You can remove it with something like "#search-menu-input:focus {
  outline: none;
}"
We can't use outline yet but can use its components.
[https://developer.mozilla.org/en-US/docs/Web/CSS/:focus](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus) explains why you shouldn't do this.

**Comment by cpeel** [comment]
> You can remove it with something like

Oh nice, thanks Ray. My inclination is to leave it as-is and not mess with browser accessibility features.

**Comment by srjfoo** [comment]
> > You can remove it with something like
> 
> Oh nice, thanks Ray. My inclination is to leave it as-is and not mess with browser accessibility features.

I'll just live with it, then.

**Comment by bpfoley** [comment]
> > You can remove it with something like
> 
> Oh nice, thanks Ray. My inclination is to leave it as-is and not mess with browser accessibility features.

Agreed. The focus ring is intended as a cue for users who use keyboard navigation -- whether it's built-in OS/browser accessibility features, or site-specific Javascript. We should keep it as is unless we've a specific reason to remove it.

**Comment by srjfoo** [comment]
> > > You can remove it with something like
> > 
> > 
> > Oh nice, thanks Ray. My inclination is to leave it as-is and not mess with browser accessibility features.
> 
> Agreed. The focus ring is intended as a cue for users who use keyboard navigation -- whether it's built-in OS/browser accessibility features, or site-specific Javascript. We should keep it as is unless we've a specific reason to remove it.

I will defer to wiser heads; I'll get used to it. I do want to mention that, at least for this search, I'm seeing a double ring in all but Safari (though I haven't looked at other browsers/OSs). Who knew that accent colors could be such a pain? Too brigh in this case, not bright enough when highlighting small pieces of text in some cases, with the same base color!

**Comment by cpeel** [comment]
Rebased to squash the commits and will merge after the CI passes.

Thanks to everyone for your great feedback!

**Pull request merged**
